toolchain/sem_ir/type_iterator.h

```diff
@@ -0,0 +1,185 @@
+  struct NonTypeValue {
+    // The type of the value.
+    SemIR::TypeId type_id;
+  };
+
+  using WorkItem = std::variant<SemIR::TypeId, SymbolicType, NonTypeValue,
+                                SemIR::SpecificInterface, EndType>;
+
+  // Get the TypeId for an instruction that is not a facet value, otherwise
+  // return SymbolicType to indicate the instruction is a symbolic facet value.
+  //
+  // If the instruction is not a type value, the return is TypeId::None.
+  //
+  // We reuse the `SymbolicType` work item here to give a nice return type.
+  auto TryGetInstIdAsTypeId(SemIR::InstId inst_id) const
+      -> std::variant<SemIR::TypeId, SymbolicType>;
+
+  // Get the instructions in the specific's instruction block as an ArrayRef.
+  auto GetSpecificArgs(SemIR::SpecificId specific_id)
+      -> llvm::ArrayRef<SemIR::InstId>;
+
+  // Push all arguments from the array into the work queue.
+  auto PushArgs(llvm::ArrayRef<SemIR::InstId> args) -> void;
+
+  // Push an instruction's type value into the work queue, or a marker if the
+  // instruction has a symbolic value.
+  auto PushInstId(SemIR::InstId inst_id) -> void;
+
+  // Push the next step into the work queue.
+  auto Push(WorkItem item) -> void;
+
+  File* sem_ir_;
+  llvm::SmallVector<WorkItem> work_list_;
+};
+
+class TypeIterator::Step {
+ public:
+  // Followed by generic parameters.
+  struct ClassStart {
+    SemIR::ClassId class_id;
+    SemIR::TypeId type_id;
+  };
+  // Followed by its fields.
+  struct StructStart {
+    SemIR::TypeId type_id;
+  };
+  // Followed by its members.
+  struct TupleStart {
+    SemIR::TypeId type_id;
+  };
+  // Followed by generic parameters.
+  struct InterfaceStart {
+    SemIR::InterfaceId interface_id;
+  };
+  // Followed by the bit width.
+  struct IntStart {
+    SemIR::TypeId type_id;
+  };
+  // Followed by the type and bound.
+  struct ArrayStart {
+    SemIR::TypeId type_id;
+  };
+  // Followed by the pointee type.
+  struct PointerStart {};
+
+  // A type value.
+  struct ConcreteType {
+    SemIR::TypeId type_id;
+  };
+  // A symbolic type value, constrained by `facet_type_id`.
+  struct SymbolicType {
+    // Either a FacetType or the TypeType singleton.
+    SemIR::TypeId facet_type_id;
+  };
+  // A symbolic template type value.
+  struct TemplateType {};
+  // A non-type value, which can be found as a generic parameter for a type.
+  struct Value {
+    // TODO: Include the type of the value somewhere.
+  };
+
+  // A *Start type, but without anything following it, so we also skip the
+  // matching `End`.
+  struct StartOnly {
+    using Any =
+        std::variant<ClassStart, StructStart, TupleStart, InterfaceStart>;
+    Any any;
+  };
+  // A *Start type, with nested steps after it, and then a matching `End`.
+  struct StartWithEnd {
+    using Any =
+        std::variant<ClassStart, StructStart, TupleStart, InterfaceStart,
+                     ArrayStart, IntStart, PointerStart>;
+    Any any;
+  };
```

Comment:
Maybe this makes me terrible, but had you considered something like:

```suggestion
  // Whereas `*Start` types normally have nested steps and a matching `End`,
  // these are only the start with no `End`.
  struct ClassStartOnly : public ClassStart {};
  struct StructStartOnly : public StructStart {};
  struct TupleStartOnly : public TupleStart {};
  struct InterfaceStartOnly : public InterfaceStart {};
```

In particular, avoiding the nested variant approach. If you actually take this advice, it might also be better to document the difference on `Step` instead of as in this suggestion.

Reply:
I had not considered it, I will give it a try against the switchified version and see how it turns out.

Reply:
I like the result of this, thanks for the idea. Sending a PR with it.